synapse/handlers/auth.py

```diff
@@ -322,6 +350,13 @@ def check_auth(
                     creds,
                     list(clientdict),
                 )
+
+                # Once the authentication flow has completed and the final
+                # operation is requested, the session should be removed so it
+                # cannot be re-used.
+                if "type" not in authdict:
```

Comment:
why is this condition required?

Reply:
Without this if-statement the session gets removed during the last stage of the UI Auth flow. With this if-statement it is removed after the operation is repeated.

An example might make this clearer:

1. Client makes request to do something, gets a 401 to do UI Auth.
2. Client chooses a flow and starts UI Auth stages.
3. Client re-makes the request from step 1, not that UI auth is done.

Without this if-statement the session gets removed in step 2n (where n is the last stage of the flow). This causes step 3 to fail since the session gets removed immediately after being fully authenticated.

I tried to clarify this in the comment, but was having some issues describing this. Please let me know what can make it clearer.

Reply:
hrm. that doesn't really match my understanding of how UI-Auth is supposed to work. AFAIK there isn't necessarily a step 3.

At step 2n, `check_auth` returns without raising, hence `validate_user_via_ui_auth` returns without raising, and the REST handler goes on to do whatever the client requested; but we haven't cleared the session because of this condition.

Possibly we need to move the call to `remove_session` up to the REST handler so that we know for certain whether the operation has completed?

Reply:
> hrm. that doesn't really match my understanding of how UI-Auth is supposed to work. AFAIK there isn't necessarily a step 3.

This was my understanding going into this as well, but doesn't seem to be how some of the tests work. I'm unsure if this is an issue with the tests or the design or something else though.

> Possibly we need to move the call to `remove_session` up to the REST handler so that we know for certain whether the operation has completed?

This would probably make it more explicit. I'll give that a try and see if it is clearer.

Reply:
> This was my understanding going into this as well, but doesn't seem to be how some of the tests work. I'm unsure if this is an issue with the tests or the design or something else though.

that sounds like an issue with the tests to me... icbw though

Reply:
My concern was that it works now and changing it might break clients. I'll look again though and make sure I understand the issue.

Reply:
The fallback tests were failing without doing this, I assumed their behavior was correct, but I noticed something odd in those tests:

https://github.com/matrix-org/synapse/blob/229eb81498b0fe1da81e9b5b333a0285acde9446/tests/rest/client/v2_alpha/test_auth.py#L101-L114

The request that includes the dummy auth actually returns 200, not 401 like I expected. I think this code is essentially re-using the session, which is what we're trying to protect from. Does that seem correct @richvdh?

Reply:
OK, so the behavior I seem to be fighting against here was added in #649 and seems to exist to fix a real bug. This makes me wonder if we really want to delete the session at all or just let them expire as they do now?

Reply:
We talked about this in #synapse-dev and came to the following conclusions:

* If the requests are idempotent we should not need to invalidate the session.
* We could invalidate the session from the REST code if they are not idempotent.
* It seems the current actions are all idempotent: deleting devices, setting a password, deactivating an account, adding a 3pid, registering, uploading a key.

The other change in here is to ensure that the same action is happening through a session (e.g. the password must be consistent when using the same session ID).

So the result is to get rid of the code that removes a "used" session.